src/engine/strat_engine/thinpool.rs

```diff
@@ -185,7 +254,69 @@ impl ThinPool {
             .map(|s| s.length)
             .sum();
 
-        Ok(data_dev_used + spare_total + meta_dev_total)
+        let mdv_total = self.mdv.segments().iter().map(|s| s.length).sum();
+
+        Ok(data_dev_used + spare_total + meta_dev_total + mdv_total)
+    }
+
+    pub fn get_filesystem_by_uuid(&mut self,
+                                  uuid: &FilesystemUuid)
+                                  -> Option<&mut StratFilesystem> {
+        self.filesystems.get_mut_by_uuid(uuid)
+    }
+
+    pub fn get_filesystem_by_name(&mut self, name: &str) -> Option<&mut StratFilesystem> {
+        self.filesystems.get_mut_by_name(name)
+    }
+
+    pub fn has_filesystems(&self) -> bool {
+        !self.filesystems.is_empty()
+    }
+
+    pub fn create_filesystem(&mut self,
+                             pool_uuid: &Uuid,
+                             fs_uuid: FilesystemUuid,
+                             name: &str,
+                             dm: &DM,
+                             size: Option<Sectors>)
+                             -> EngineResult<()> {
+        let new_filesystem =
+            try!(StratFilesystem::initialize(pool_uuid, fs_uuid, name, dm, self, size));
```

Comment:
deferring this to another PR, this results in `StratFilesystem::setup` and `initialize` having much less functionality.